### PR TITLE
Create more scalable revision graph (performance)

### LINKED ISSUE #5593: Commit graph continuity issue in 3.0
**Do you want to request a *feature* or report a *bug*?**
Bug

A view of GitExtensions 2.5 and 3.0 side by side
image

Notice the discontinuity in the graph line on the right.
Unfortunately this is a private repo and I can't share it. This is the first time I am seeing this so it's hard to repro. I am creating this issue to inform you of the investigation progress trying to find the root cause.

Turning on Reflog References in 3.0 fixes this. I'll keep you posted.

### PR DESCRIPTION
**Pull request by spdr870**
Fixes #5593 and #5167

Changes proposed in this pull request:
- The revision graph is rewritten to work independent of ordering of the revisions
- Memory usage decreased (linux repo fully loaded dropped from 6.8Gb to 1.3Gb)
- Performance of loading the graph (not the commits!) is significantly faster in larger repos
- Performance of loading the commits (not the graph!) is slower (about 2 sec for loading the linux repo)
- Because the graph will now be visible in larger repositories, the slow scrolling becomes more obvious. Before, the graph was usually only visible for the top x commits, because it took forever to load.

Screenshots before and after (if PR changes UI):
- Not much changed in the ui
- Ordering of the lanes/revisions changed a little
Left = new, Old = current
image

What did I do to test the code and ensure quality:
- Manual testing on various repositories

What do I propose to change:
- [x]  Remove setting for --date-order and --topo-order. It doesn't do anything.
- [x]  I broke (deleted) the TryGetTooltip code.... 
- [x]  There are some graphical glitches in the top op the graph when scrolling
- [x]  Fix change the relative highlight to another branch with ctrl-shift-b
- [x] Add some unit tests

Has been tested on (remove any that don't apply):
- GIT 2.19
- Windows 10

### CODE REVIEW

**Comment by mdonatas** [comment]
Really nice progress in such a short time!
I've tried this PR on my work repo and it seems to work fine. Would need to run non debug build to evaluate performance.

File: GitUI/UserControls/RevisionGrid/Graph/RevisionGraphRevision.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        public bool HasRef { get; set; }
+        public bool IsCheckedOut { get; set; }
+
+        public int Score { get; private set; }
+
+        public int LaneColor { get; set; }
+
+        public int IncreaseScore(int delta)
+        {
+            if (delta + 1 > Score)
```
Comment (RussKie) [inline GitUI/UserControls/RevisionGrid/Graph/RevisionGraphRevision.cs]: invert to reduce nesting

File: GitUI/UserControls/RevisionGrid/Graph/RevisionGraphRevision.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        public ObjectId Objectid { get; set; }
+
+        public ConcurrentBag<RevisionGraphRevision> Parents { get; private set; }
+        public ConcurrentBag<RevisionGraphRevision> Children { get; private set; }
+        public SynchronizedCollection<RevisionGraphSegment> StartSegments { get; private set; }
+        public ConcurrentBag<RevisionGraphSegment> EndSegments { get; private set; }
+
+        public void MakeRelative()
+        {
+            if (!IsRelative)
```
Comment (RussKie) [inline GitUI/UserControls/RevisionGrid/Graph/RevisionGraphRevision.cs]: invert to reduce nesting

File: GitUI/UserControls/RevisionGrid/Graph/RevisionGraphRow.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        public SynchronizedCollection<RevisionGraphSegment> Segments { get; private set; }
+
+        private IReadOnlyDictionary<RevisionGraphSegment, int> _segmentLanes;
+
+        private int _laneCount;
+
+        private void BuildSegmentLanes()
+        {
+            int currentRevisionLane = -1;
+            if (_segmentLanes == null)
```
Comment (RussKie) [inline GitUI/UserControls/RevisionGrid/Graph/RevisionGraphRow.cs]: invert to reduce nesting

**Comment by spdr870** [comment]
@RussKie I added some tests. It is now pretty easy to add more.

About the complexity. True, it is complex. For me it is easier then the current code, but also because I wrote this myself. I will add some documentation. At least what the purpose is of each method and class. The tests might help a little.

**Comment by spdr870** [comment]
I did some testing in more scenario's. Did some fixes:
- Repository with just 1 commit was not rendering
- Repository with just 1 commit crashed after refresh
- Last commit not drawn in graph
- Refreshing in fast (scripted in brute force) could cause crash

Also did some minor performance improvements in initial commit loading. This PR is ready to be reviewed and tested.

**Comment by gerhardol** [comment]
This works well. No major comments on the source. This is complex but so was the previous implementation.
Of course a risk to change this just before the 3.0 RC

The following is not a blocker but WorkTree/CommitIndex/HEAD should be kept together
image

**Comment by gerhardol** [comment]
Similar for stash (untracked files) in the screenshot I realize now

I do not know when the (minor) visual problem occurs

**Comment by spdr870** [comment]
> This works well. No major comments on the source. This is complex but so was the previous implementation.
> Of course a risk to change this just before the 3.0 RC
> 
> The following is not a blocker but WorkTree/CommitIndex/HEAD should be kept together
> image

I might have already fixed this in this commit b57c071d6a35fd5a08e0f05bb5cb90fb3cc9a496. What commit did you noticed this behaviour? I could simulate before this commit:
image
After b57c071d6a35fd5a08e0f05bb5cb90fb3cc9a496:
image

**Comment by spdr870** [comment]
@RussKie WIP is still pending. How to remove the WIP status? I did change the title of this PR.

**Comment by vbjay** [comment]
In my email.

A GitHub App, WIP, installed on your account is requesting updated
permissions.

Visit https://github.com/settings/installations/99232/permissions/update to
review and accept or deny these permissions.

You can view pending requests directly by visiting your account’s settings
page and clicking on Applications. If you run into problems, please contact
support by visiting https://github.com/contact.

Thanks,
Your friends at GitHub


So on gitextensions I'm sure you need to do the same.

**Comment by spdr870** [comment]
@vbjay Thanks! I changed the permissions.

**Comment by vbjay** [comment]
Might need to flip the title to get WIP to react or close and reopen.

**Comment by mdonatas** [comment]
A day or two ago I've noticed regression in how colors of merge segments are chosen. I'll test later to add screenshots. Just saying this now so you wouldn't rush to merge these changes :)

**Comment by spdr870** [comment]
@mdonatas I noticed this as well. And because the graphics of the graph are very important, I worked on this a bit. Hope you like the changes!

- I try to detect the 'main' branches. These main branches keep there color. Left = new.
image
- I changed the color scheme. 
- When a branch is created and then merged, only this part is colored. The main branch keeps its color.
- Related branches are thicker then unrelated branches.
image
- The HEAD is preferable orange. This makes it recognizable. This orange is a 'softer' version of the current red.
image

If you have any thought/suggestions let me know! I am not a graphics expert, open to all suggestions.

**Comment by spdr870** [comment]
An example of the branch detecting. Notice in the left screenschot that the main branch is kept on the left. The feature branches are on the right. This is clearly visible when you look at the merge commits. The right screenshot is the same repository in the 3.00. Notice that the merge commits are in different lanes. Unfortunately the coloring isn't perfect. But still, I think it is an improvement over the current graph.
image

**Comment by mdonatas** [comment]
Really love this idea to try and keep main branch on the left, looks marvelous! And the merge span is also fixed, awesome!

At least personally I prefer previously used color scheme. To me it looks more modern and crisp. The new color scheme looks pastel and closer to colors used at my kids kindergarten :D

As for the lane thickness of the main branch it looks a little bit childish and fills too much of the white-space when there are many branches thus I prefer the old way of distinguishing by color alone.
image

But that's just my opinion. I'd like to hear what others think.

p.s. there is a single pixel row missing in the middle of the commit row, it's visible in the screenshot if you look closely. I have a 125% scaling on my laptop if that's relevant.

**Comment by RussKie** [comment]
I like the idea of the currently checked out branch to be on the left.
I don't mind the new colour scheme, but don't have a strong opinion on either.

I am a little confused by the @mdonatas's screenshot - why are all branches coloured? Shouldn't only the currently checked out branch be coloured and all other ones be greyed out?

I presume the branches with thicker lines are all merged into the currently checked out branch.

**Comment by spdr870** [comment]
@mdonatas I will fix this small gap between the rows. Probably a rounding issue.

About your opinion on the styling; I really suck at graphical design. I would be nice if someone with the proper skills could help with this. If others agree, I have no problems with rolling back the color/with changes. My main focus is the graph itself and its topology.

@RussKie Its a setting..:) In the context menu view->draw non relatives gray.

**Comment by RussKie** [comment]
Ah thanks. We have too many settings 😄

**Comment by mdonatas** [comment]
In case we leave "Draw non relatives gray" setting could we use the same thickness for all rows?

**Comment by spdr870** [comment]
> In case we leave "Draw non relatives gray" setting could we use the same thickness for all rows?

Yes... it makes not sense to change the thickness then. Just forgot about the setting for a moment...:(

**Comment by gerhardol** [comment]
> I might have already fixed this in this commit b57c071. In what commit did you noticed this behaviour? I could simulate before this commit:

This was with 04e1b9aac97d66dc1292633f5805f2a06c28988c, so this should be fixed.

I like the thickness and lane changes but prefer the old colors

**Comment by spdr870** [comment]
I reverted the styling changes. If we want to change anything, lets separate it from this PR. The lane gaps are fixed. I could copy a fix from the old code.

**Comment by mdonatas** [comment]
I sometimes notice lines becoming thicker (left-most gray line becomes visually thicker, open full size image to see it):
image

When zommed in it's evident why:
image

This doesn't happen with master version.

**Comment by spdr870** [comment]
> I sometimes notice lines becoming thicker (left-most gray line becomes visually thicker, open full size image to see it):

Good find! It is fixed. All lines where drawn as bezier using antialias. Straight lines should be sharp lines, draws as a line without antialias.

**Comment by drewnoakes** [comment]
When I removed the edges from the lines in #5087 I worked with someone else (apologies I have forgotten whom) and we made the lines render without antialising. This then required care to avoid jagged transitions between the bezier curve and the straight section. It's possible however, and you can use the old code as a reference.

**Comment by spdr870** [comment]
> When I removed the edges from the lines in #5087 I worked with someone else (apologies I have forgotten whom) and we made the lines render without antialising. This then required care to avoid jagged transitions between the bezier curve and the straight section. It's possible however, and you can use the old code as a reference.

Are you referring to this code? I copied most of the old render code.
`                    // Anti-aliasing seems to introduce an offset of two thirds
                    // of a pixel to the right - compensate it.
                    g.SmoothingMode = SmoothingMode.AntiAlias;
                    float offset = -0.667F;`

**Comment by drewnoakes** [comment]
> Are you referring to this code? 

Yep. That screenshot looks great!

**Comment by spdr870** [comment]
> Wow, the graph for the Linux repo is loaded such quickly, now!
> 
> Is it intended that side-by-side graph lanes have the identical color?

We have just 8 colors. There is no check that colors in lanes next to each other are the same. Because of this, we have color collisions. I hope this is not a big problem. Nice to see that you notice the performance difference! For the repository I work on, this means that I finally see the graph for the older commits. Normally I never see commits over 3 months old. Now I even see the graph for commits over 8 years old (repo is 150.000 commits).

**Comment by RussKie** [comment]
Is there a technical limitation presenting us from having more than 8
colours?

**Comment by spdr870** [comment]
Nope, we can have as many colors as we want.

**Comment by spdr870** [comment]
Created some basic documentation on the wiki:
https://github.com/gitextensions/gitextensions/wiki/Revision-Graph

**Comment by RussKie** [comment]
I'm very short on time to properly review.
Let's merge it, so we can publish beta2.

**Pull request merged**